### ISSUE TITLE
IO::Select::_fileno(): Ambiguous code, can generate uninitialized value warning

### ISSUE THREAD

**Issue by jkeenan** [Perl/perl5#19039]
There is code in subroutine `_fileno()` in `dist/IO/lib/IO/Select.pm` that makes an imprecise comparison and may generate a `Use of uninitialized value in numeric eq (==)` warning.
```
 55 sub _fileno
 56 {
 57  my($self, $f) = @_;
 58  return unless defined $f;
 59  $f = $f->[0] if ref($f) eq 'ARRAY';
 60  if($f =~ /^[0-9]+$/) { # plain file number
 61   return $f;
 62  }
 63  elsif(defined(my $fd = fileno($f))) {
 64   return $fd;
 65  }
 66  else {
 67   # Neither a plain file number nor an opened filehandle; but maybe it was
 68   # previously registered and has since been closed. ->remove still wants to
 69   # know what fileno it had
 70   foreach my $i ( FIRST_FD .. $#$self ) {
 71    return $i - FIRST_FD if $self->[$i] == $f;
 72   }
 73   return undef;
 74  }
 75 }
```
At line 71, there is no guarantee that `$self->[$i]` holds a numeric value suitable for comparison with the `==` operator.  Nor is there any guarantee that `$self->[$i]` even holds a defined value.

In this branch [jkeenan/io-select-20210809](https://github.com/jkeenan/perl5/tree/io-select-20210809), I have [hacked up IO::Select](https://github.com/jkeenan/perl5/blob/fd9580b3e6cb5345fe12cb93f8eefd955a13a35b/dist/IO/lib/IO/Select.pm#L55) with some debugging code.
```
$ git diff blead..io-select-20210809 -- dist/IO/lib/IO/Select.pm
diff --git a/dist/IO/lib/IO/Select.pm b/dist/IO/lib/IO/Select.pm
index d44889e54a..4fb7ebd4ee 100644
--- a/dist/IO/lib/IO/Select.pm
+++ b/dist/IO/lib/IO/Select.pm
@@ -67,8 +67,15 @@ sub _fileno
   # Neither a plain file number nor an opened filehandle; but maybe it was
   # previously registered and has since been closed. ->remove still wants to
   # know what fileno it had
+  use warnings;
+  print STDERR "AAA: range: first_fd: ", FIRST_FD, "\tmax index: $#$self\n";
+  print STDERR "AAA: \$f is ", (defined($f) ? "defined" : "not defined"), "\n";
   foreach my $i ( FIRST_FD .. $#$self ) {
-   return $i - FIRST_FD if $self->[$i] == $f;
+    print STDERR "BBB: \$self->[$i] is ",
+        (defined($self->[$i]) ? "defined as $self->[$i]" : "not defined"), "\n";
+   if ($self->[$i] == $f) {
+    return $i - FIRST_FD;
+   }
   }
   return undef;
  }
```
When you build perl on this branch and run the test suite, you will see that this code is exercised in two test files, `dist/IO/t/gh17447.t` (which is coded with `use warnings`) and `dist/IO/t/io_sel.t` (which is *not* coded with `use warnings`).  Output:
```
../dist/IO/t/gh17447.t .. 
1..2
AAA: range: first_fd: 2	max index: 5
AAA: $f is defined
BBB: $self->[2] is not defined
Use of uninitialized value in numeric eq (==) at ../../lib/IO/Select.pm line 76.
BBB: $self->[3] is not defined
Use of uninitialized value in numeric eq (==) at ../../lib/IO/Select.pm line 76.
BBB: $self->[4] is not defined
Use of uninitialized value in numeric eq (==) at ../../lib/IO/Select.pm line 76.
BBB: $self->[5] is defined as GLOB(0x800adcc90)
ok 1 - $select has 1 handle
ok 2 - $select has 0 handles
ok
../dist/IO/t/io_sel.t ... 
1..27
ok 1
...
ok 23
AAA: range: first_fd: 2	max index: 5
AAA: $f is defined
BBB: $self->[2] is defined as GLOB(0x800adcf00)
BBB: $self->[3] is not defined
BBB: $self->[4] is not defined
BBB: $self->[5] is defined as IO::Socket::INET=GLOB(0x800bb5330)
ok 24 - added socket
ok 25 - removed closed socket
ok 26 - count() updated
ok 27 - bits() updated
ok
All tests successful.
Files=2, Tests=29,  0 wallclock secs ( 0.01 usr  0.00 sys +  0.07 cusr  0.02 csys =  0.09 CPU)
Result: PASS
/home/jkeenan/gitwork/perl
```
`dist/IO/lib/IO/Select.pm` itself is coded with `use warnings::register;`.

I encountered the uninitialized value warning while running the `o conf init test_report` command when initializing CPAN::Reporter.

I suspect that `_fileno` returns an undefined value when `$self->[$i]` is undefined -- but the intent of the code at line 71 is ambiguous.  This code was committed in d20228ab36e in January 2020.  @toddr or anyone, can you take a look?

Thank you very much.
Jim Keenan

**Comment by toddr** [comment]
It appears the patch came from @leonerd if you dive into the RT case.

https://rt.cpan.org/Public/Bug/Display.html?id=77740

Which eventually became  https://github.com/Perl/perl5/issues/17447

**Comment by toddr** [comment]
@jkeenan I think this was fixed via cab7765183cdc58e214067bafc145253606c347d

**Comment by jkeenan** [comment]
> @jkeenan I think this was fixed via [cab7765](https://github.com/Perl/perl5/commit/cab7765183cdc58e214067bafc145253606c347d)

Yes, it does appear to be fixed.  I rebased my branch on blead and adjusted the positioning of the `BBB` debugging code:
```
diff --git a/dist/IO/lib/IO/Select.pm b/dist/IO/lib/IO/Select.pm
index 325ef22273..823313729e 100644
--- a/dist/IO/lib/IO/Select.pm
+++ b/dist/IO/lib/IO/Select.pm
@@ -67,7 +67,12 @@ sub _fileno
   # Neither a plain file number nor an opened filehandle; but maybe it was
   # previously registered and has since been closed. ->remove still wants to
   # know what fileno it had
+  use warnings;
+  print STDERR "AAA: range: first_fd: ", FIRST_FD, "\tmax index: $#$self\n";
+  print STDERR "AAA: \$f is ", (defined($f) ? "defined" : "not defined"), "\n";
   foreach my $i ( FIRST_FD .. $#$self ) {
+    print STDERR "BBB: \$self->[$i] is ",
+        (defined($self->[$i]) ? "defined as $self->[$i]" : "not defined"), "\n";
    return $i - FIRST_FD if defined $self->[$i] && $self->[$i] == $f;
   }
   return undef;
```
I then ran:
```
$ cd t; TEST_JOBS=1 ./perl harness ../dist/IO/t/gh17447.t ../dist/IO/t/io_sel.t;
cd -
../dist/IO/t/gh17447.t .. AAA: range: first_fd: 2	max index: 5
AAA: $f is defined
BBB: $self->[2] is not defined
BBB: $self->[3] is not defined
BBB: $self->[4] is not defined
BBB: $self->[5] is defined as GLOB(0x800aeac90)
../dist/IO/t/gh17447.t .. ok   
../dist/IO/t/io_sel.t ... 1/27 AAA: range: first_fd: 2	max index: 5
AAA: $f is defined
BBB: $self->[2] is defined as GLOB(0x800aeaf00)
BBB: $self->[3] is not defined
BBB: $self->[4] is not defined
BBB: $self->[5] is defined as IO::Socket::INET=GLOB(0x800bca228)
../dist/IO/t/io_sel.t ... ok     
All tests successful.
Files=2, Tests=29,  1 wallclock secs ( 0.02 usr  0.00 sys +  0.12 cusr  0.01 csys =  0.15 CPU)
Result: PASS
```
The debugging code is emitted but no warnings are emitted.  Closing.

Thank you very much.
Jim Keenan
